design/20190708.certificate-request-crd.md

```diff
@@ -0,0 +1,242 @@
+}
+
+// CertificateStatus defines the observed state of CertificateRequest and
+// resulting signed certificate.
+type CertificateRequestStatus struct {
+	// +optional
+	Conditions []CertificateRequestCondition `json:"conditions,omitempty"`
+
+	// Byte slice containing a PEM encoded signed certificate resulting from the
+	// given certificate signing request.
+	// +optional
+	Certificate []byte `json:"certificate,omitempty"`
+
+	// Byte slice containing the PEM encoded certificate authority of the signed
+	// certificate.
+	// +optional
+	CA []byte `json:"ca,omitempty"`
+}
+```
+
+The `CertificateRequestCondition` resembles much the same of the
+`CertificateRequestCondition`.
+
+The `ObjectReference` field type has had a new field `Group` added as follows:
+
+```
+// ObjectReference is a reference to an object with a given name, kind and group.
+type ObjectReference struct {
+	Name string `json:"name"`
+	// +optional
+	Kind string `json:"kind,omitempty"`
+	// +optional
+	Group string `json:"group,omitempty"`
+}
+```
+
+The group refers to the API group that the target Issuer belongs to. This
+enables namespacing of references to different issuers of external API groups.
+
+### Controller Behaviour
+
+The philociphy for the `CertificateRequest` controllers are planned to be as
+minimal as possible in that the single goal of them is to enable it's owning
+`Issuer` to create the resulting certificate. Once a sync on a
+`CertificateRequest` has been observed, the general flow is as follows:
+
+- Check the group belongs to the owning `Issuer`, exit if not.
+- Check if `CertificateRequest` is in a failed state, exit if true.
+- Check the `Issuer` type is of the same type, exit if not.
+- Verify the Spec of the `CertificateRequest`.
+- If a certificate exits then update the status if needed and exit.
+- Sign the certificate via the Issuer using the contents of Spec.
+
+It is worth noting that whether the certificate is invalid, out-of-date or
+failed then the controller should take no further action on the resource. It is
+the responsibility of a higher level controller such as the `Certificate`
+controller to take further action to retry the certificate issuance through
+managing the life cycle of the `CertificateRequest` resources.
+
+With all `Issuer`s updated with `CertificateRequest` controllers, the
+`Certificate` controller will be migrated to begin to use and manage the life
+cycle `CertificateRequest`s to resolve it's Spec. Further concrete
+implementation details TBD.
+
+### Internal API Resource Behaviour
+
+The group name of `IssuerRef` inside `CertificateRequest`s are to be defaulted
+to "certmanager.k8s.io" if the field is empty, using a mutating webhook. This
+means that if unspecified, `CertificateRequest` objects will be put into the
+ownership of the default pool of issuers in the cert-manager project.
+
+### Test Plan
+
+Standard unit and end-to-end tests will be used to verify new behaviour, as used
+by cert-manager currently.
```

Comment:
Maybe note that existing e2e tests for Certificate resources will also provide a good signal once we swap the Certificate controller implementation over